src/chain/shared_state.rs

```diff
@@ -309,6 +307,17 @@ impl SectionProofChain {
         }
         true
     }
+
+    pub fn last_public_key(&self) -> &BlsPublicKey {
+        self.blocks
+            .last()
+            .map(|block| &block.key)
+            .unwrap_or(&self.genesis_pk)
+    }
+
+    pub fn all_keys(&self) -> impl Iterator<Item = &BlsPublicKey> {
+        iter::once(&self.genesis_pk).chain(self.blocks.iter().map(|block| &block.key))
```

Comment:
will this have many duplicated key among blocks?

Reply:
Probably not. See above, these are BLS keys, and every one of such keys corresponds to a specific set of members of a section.
(They are currently emulated using regular keys, but this is the way it's supposed to be in the future.)